sys/include/usb/usbus.h

```diff
@@ -0,0 +1,515 @@
+#define USBUS_MSG_TYPE_HANDLER      (0x0700) /**< Generic handler request */
+/** @} */
+
+/**
+ * @name USBUS handler subscription flags
+ *
+ * @{
+ */
+#define USBUS_HANDLER_FLAG_RESET    (0x0001) /**< Report reset event */
+#define USBUS_HANDLER_FLAG_SOF      (0x0002) /**< Report SOF events */
+#define USBUS_HANDLER_FLAG_SUSPEND  (0x0004) /**< Report suspend events */
+#define USBUS_HANDLER_FLAG_RESUME   (0x0008) /**< Report resume from suspend */
+#define USBUS_HANDLER_FLAG_TR_FAIL  (0x0010) /**< Report transfer fail */
+#define USBUS_HANDLER_FLAG_TR_STALL (0x0020) /**< Report transfer stall complete */
+/** @} */
+
+/**
+ * @brief state machine states for the global USBUS thread
+ */
+typedef enum {
+    USBUS_STATE_DISCONNECT, /**< Device is disconnected from the host */
+    USBUS_STATE_RESET,      /**< Reset condition received */
+    USBUS_STATE_ADDR,       /**< Address configured */
+    USBUS_STATE_CONFIGURED, /**< Peripheral is configured */
+    USBUS_STATE_SUSPEND,    /**< Peripheral is suspended by the host */
+} usbus_state_t;
+
+/**
+ * @brief USBUS setup request state machine
+ */
+typedef enum {
+    USBUS_SETUPRQ_READY,      /**< Ready for new setup request */
+    USBUS_SETUPRQ_INDATA,     /**< Request received with expected DATA IN stage */
+    USBUS_SETUPRQ_OUTACK,     /**< Expecting a zero-length ack out from host */
+    USBUS_SETUPRQ_OUTDATA,    /**< Data out expected */
+    USBUS_SETUPRQ_INACK,      /**< Expecting a zero-length ack in request
+                                   from the host */
+} usbus_setuprq_state_t;
+
+/**
+ * @brief USBUS string type
+ */
+typedef struct usbus_string {
+    struct usbus_string *next;  /**< Ptr to the next registered string */
+    const char *str;            /**< C string to use as content        */
+    uint16_t idx;               /**< USB string index                  */
+} usbus_string_t;
+
+/**
+ * @brief USBUS context forward declaration
+ */
+typedef struct usbus usbus_t;
+
+/**
+ * @brief USBUS event handler forward declaration
+ */
+typedef struct usbus_handler usbus_handler_t;
+
+/**
+ * @brief USBUS generic header generator
+ *
+ * @param usbus The usbus context
+ * @param arg   Additional argument for the header generator
+ *
+ * @return      Length of the generated header
+ */
+typedef size_t (*gen_hdr)(usbus_t *usbus, void *arg);
+
+/**
+ * @brief USBUS generic header generator generated length
+ *
+ * Must return the length of the header that will be generated by @ref gen_hdr
+ *
+ * @param usbus The usbus context
+ * @param arg   Additional argument for the header generator
+ *
+ * @return      Length of the generated header
+ */
+typedef size_t (*hdr_len)(usbus_t *usbus, void *arg);
+
+/**
+ * @brief USBUS header generator function pointers
+ */
+typedef struct {
+    gen_hdr get_header;         /**< function ptr to retrieve the header */
+    hdr_len get_header_len;     /**< function ptr to retrieve the header len */
+} usbus_hdr_gen_funcs_t;
```

Comment:
Experience has learned that most headers have a fixed size. Changing this to:
```
typedef struct {
    gen_hdr get_header;         /**< function ptr to retrieve the header */
    union {
        hdr_len get_header_len;     /**< function ptr to retrieve the header len */
        size_t fixed;               /**< Fixed length of the header */
    } len;
    usbus_hdr_len_t len_type     /**< Either USBUS_HDR_LEN_FIXED or USBUS_HDR_LEN_FUNC */
} usbus_hdr_gen_funcs_t;
```
Allows for both fixed size and function based at almost zero cost (current function calls for the length are 4B of flash)